src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs

```diff
@@ -13,44 +14,108 @@ namespace Microsoft.CodeAnalysis
+
+        private readonly object _collectionLock = new object();
 
         internal AdditionalSourcesCollection()
         {
-            _sourcesAdded = PooledDictionary<string, SourceText>.GetInstance();
+            _hintNames = PooledHashSet<string>.GetInstance();
+            _sourcesAdded = ArrayBuilder<GeneratedSourceText>.GetInstance();
         }
 
-        internal AdditionalSourcesCollection(ImmutableArray<GeneratedSourceText> sources)
+        internal AdditionalSourcesCollection(ImmutableArray<GeneratedSourceText> existingSources)
             : this()
         {
-            foreach (var source in sources)
-            {
-                _sourcesAdded.Add(source.HintName, source.Text);
-            }
+            _sourcesAdded.AddRange(existingSources);
         }
 
         public void Add(string hintName, SourceText source)
         {
-            _sourcesAdded.Add(hintName, source);
+            if (string.IsNullOrWhiteSpace(hintName))
+            {
+                throw new ArgumentNullException(nameof(hintName));
+            }
+
+            // allow any identifier character or [.,-_ ()[]{}]
+            for (int i = 0; i < hintName.Length; i++)
+            {
+                char c = hintName[i];
+                if (!UnicodeCharacterUtilities.IsIdentifierPartCharacter(c)
+                    && c != '.'
+                    && c != ','
+                    && c != '-'
+                    && c != '_'
+                    && c != ' '
+                    && c != '('
+                    && c != ')'
+                    && c != '['
+                    && c != ']'
+                    && c != '{'
+                    && c != '}')
+                {
+                    throw new ArgumentException(string.Format(CodeAnalysisResources.HintNameInvalidChar, c, i), nameof(hintName));
+                }
+            }
+
+            hintName = AppendExtensionIfRequired(hintName);
+            lock (_collectionLock)
+            {
+                if (_hintNames.Contains(hintName))
+                {
+                    throw new ArgumentException(CodeAnalysisResources.HintNameUniquePerGenerator, nameof(hintName));
+                }
+
+                _hintNames.Add(hintName);
+                _sourcesAdded.Add(new GeneratedSourceText(hintName, source));
+            }
         }
 
         public void RemoveSource(string hintName)
         {
-            _sourcesAdded.Remove(hintName);
+            hintName = AppendExtensionIfRequired(hintName);
+            lock (_collectionLock)
+            {
+                if (_hintNames.Contains(hintName))
+                {
+                    _hintNames.Remove(hintName);
+                    for (int i = 0; i < _sourcesAdded.Count; i++)
+                    {
+                        // check the hashcodes, as thats the comparison we're using to put the names into _hintNames
+                        if (_sourcesAdded[i].HintName.GetHashCode() == hintName.GetHashCode())
```

Comment:
Can't compare `HashCode` here because you can get false matches.

Reply:
But we're comparing the hash of the string, it'll always be the same if they're considered equal, no? Otherwise the HashSet<string> wouldn't work either?

---
In reply to: [435565433](https://github.com/dotnet/roslyn/pull/44768#discussion_r435565433) [](ancestors = 435565433)